### ISSUE TITLE
dune fails to build on termux due to missing wait3/pthreads

### ISSUE THREAD

**Issue by seanbreckenridge** [ocaml/dune#8676]
## Expected Behavior

`opam install dune` installs `dune`

Perhaps related #7385?

## Actual Behavior

```
$ opam install dune
The following actions will be performed:
  ∗ install dune 3.10.0

<><> Gathering sources ><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
[dune.3.10.0] found in cache

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
[ERROR] The compilation of dune failed at
        "/data/data/com.termux/files/home/.opam/5.0.0/bin/ocaml boot/bootstrap.ml
        -j 1".

#=== ERROR while compiling dune.3.10.0 ========================================#
# context     2.0.4 | linux/arm64 | ocaml-base-compiler.5.0.0 | https://opam.ocaml.org#b61304c6
# path        ~/.opam/5.0.0/.opam-switch/build/dune.3.10.0
# command     ~/.opam/5.0.0/bin/ocaml boot/bootstrap.ml -j 1
# exit-code   2
# env-file    ~/.opam/log/dune-18226-d2c37b.env
# output-file ~/.opam/log/dune-18226-d2c37b.out
### output ###
# extern value caml_set_oo_id();
# [...]
# /data/data/com.termux/files/home/.opam/5.0.0/lib/ocaml/caml/mlvalues.h:423:18: note: conflicting prototype is here
# CAMLextern value caml_set_oo_id(value obj);
#                  ^
# 2 warnings generated.
# ./.duneboot.exe -j 1
# cd _boot && /data/data/com.termux/files/home/.opam/5.0.0/bin/ocamlopt.opt -o ../_boot/dune.exe -g -I +unix -I +threads unix.cmxa threads.cmxa -no-alias-deps -w -49-6 readdir.c wait3_stubs.c signal_stubs.c platform_stubs.c copyfile_stubs.c dune_digest_stubs.c xdg_stubs.c spawn_stubs.c dune_stats_stubs.c dune_flock.c csexp_rpc_stubs.c fsevents_stubs.c inotify_stubs.c fswatch_win_stubs.c sha512_[...]
# otherlibs/stdune/src/wait3_stubs.c:59:9: error: call to undeclared function 'wait3'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
#   pid = wait3(&status, cv_flags, &ru);
#         ^
# 1 error generated.
#



<><> Error report <><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
┌─ The following actions failed
│ λ build dune 3.10.0
```

I tried cloning the repository and running `make release`/`configure`, but I get the same error

I'd guess that `wait3` is something that is typically available on linux but android does not have it (not sure if I need to install a package?). I looked through the code and I see windows doesn't have access to it either

## Reproduction

Because I had previously tried to build `opam` from source, I did install `pkg install rsync clang coreutils dash diffutils make nc`

If its related, documented how I installed opam [here](https://exobrain.sean.fish/devlog/ocaml_android/), but TL;DR is through a pre-compiled binary for termux, then disabling the bubblewrap sandbox by editing `~/.config/opam` and installing `5.0.0` with `LDFLAGS="-landroid-shmem" opam switch create 5.0.0 --jobs=1`, mostly by following [instructions from here](https://github.com/ocaml/opam-repository/issues/22748)

I did find this [SBCL thread](https://bugs.launchpad.net/sbcl/+bug/1856377) that seems to be discussing the same issue.

## Specifications

- Version of `dune` (output of `dune --version`): --
- Version of `ocaml` (output of `ocamlc --version`): 5.0.0
- Verison of `opam`: 2.0.4
- Operating system (distribution and version): termux on android on a Pixel 5a: `Linux localhost 4.19.269-g8728c337137c-ab10161573 #1 SMP PREEMPT Thu May 18 11:19:41 UTC 2023 aarch64 Android`

## Additional information 

Link to `opam install dune --verbose` (though, is not too different): https://gist.github.com/seanbreckenridge/4e86d5a2b74aa4104fa129b15a1eef82

**Comment by romanovj** [comment]
you should wrap the function

https://github.com/termux/termux-packages/blob/489d245fe5fca6621560dbc7fdba354bc9dc9095/packages/dash/wait3.patch#L8

**Comment by Alizter** [comment]
I had a similar issue trying to build on Haiku. https://github.com/ocaml/dune/issues/8551

Currently we have only these consumers of the stub:
```ocaml

  let wait_win32 t =
    while not (wait_nonblocking_win32 t) do
      Mutex.unlock t.mutex;
      Thread.delay 0.001;
      Mutex.lock t.mutex
    done
  ;;

  let wait_unix t =
    Mutex.unlock t.mutex;
    let proc_info = Proc.wait [] in
    Mutex.lock t.mutex;
    Process_table.remove t proc_info
  ;;

  let wait = if Sys.win32 then wait_win32 else wait_unix

```
We could generalize the support here to use Thread.delay when wait3 is not supported.

cc @rgrinberg any further comments?

**Comment by Alizter** [comment]
Actually reading around it seems that wait3 is actually deprecated and that wait4 is more something we should use:

https://www.gnu.org/software/libc/manual/html_node/BSD-Wait-Functions.html

**Comment by seanbreckenridge** [comment]
Applied:

```diff
diff --git a/otherlibs/stdune/src/wait3_stubs.c b/otherlibs/stdune/src/wait3_stubs.c
index f84c9288d..edab23e02 100644
--- a/otherlibs/stdune/src/wait3_stubs.c
+++ b/otherlibs/stdune/src/wait3_stubs.c
@@ -45,6 +45,10 @@ static value alloc_process_status(int status) {
 
 static int wait_flag_table[] = {WNOHANG, WUNTRACED};
 
+pid_t wait3(int *status, int options, struct rusage *rusage) {
+  return wait4(-1, status, options, rusage);
+}
+
 value dune_wait3(value flags) {
   CAMLparam1(flags);
   CAMLlocal2(times, res);
```

And that does fix that part of the compilation, but it fails because it doesn't have access to `pthreads`.

This is probably the same issue as https://github.com/ocaml/dune/issues/7385 as I'm running into the same phread issues they described, though that [said to open if there were issues building](https://github.com/ocaml/dune/issues/7385#issuecomment-1480587252)

After cloning and applying that patch, I ran  `make release`

```
$ make release
ocamlc -output-complete-exe -w -24 -g -o .duneboot.exe -I boot -I +unix unix.cma boot/libs.ml boot/duneboot.ml
/data/data/com.termux/files/usr/tmp/camlobjd8f0ee.c:17174:14: warning: a function declaration without a prototype is deprecated in all versions of C and is treated as a zero-parameter prototype in C2x, conflicting with a previous declaration [-Wdeprecated-non-prototype]
extern value caml_get_public_method();
             ^
/data/data/com.termux/files/home/.opam/5.0.0/lib/ocaml/caml/mlvalues.h:236:18: note: conflicting prototype is here
CAMLextern value caml_get_public_method (value obj, value tag);
                 ^
/data/data/com.termux/files/usr/tmp/camlobjd8f0ee.c:17376:14: warning: a function declaration without a prototype is deprecated in all versions of C and is treated as a zero-parameter prototype in C2x, conflicting with a previous declaration [-Wdeprecated-non-prototype]
extern value caml_set_oo_id();
             ^
/data/data/com.termux/files/home/.opam/5.0.0/lib/ocaml/caml/mlvalues.h:423:18: note: conflicting prototype is here
CAMLextern value caml_set_oo_id(value obj);
                 ^
2 warnings generated.
./.duneboot.exe
Done: 145/1347 (jobs: 8)cd _boot && /data/data/com.termux/files/home/.opam/5.0.0/bin/ocamlopt.opt -c -g -I +unix -I +threads spawn_stubs.c
vendor/spawn/src/spawn_stubs.c:706:3: error: call to undeclared function 'pthread_setcancelstate'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &cancel_state);
  ^
vendor/spawn/src/spawn_stubs.c:706:26: error: use of undeclared identifier 'PTHREAD_CANCEL_DISABLE'
  pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &cancel_state);
                         ^
2 errors generated.
make: *** [Makefile:53: _boot/dune.exe] Error 2
```

Seemingly because threads cant be cancelled on android...

Found the same thing being disscussed here: https://github.com/axel-download-accelerator/axel/issues/150

where they replaced it with `pthread_kill`

and here: https://sourceforge.net/p/djvu/bugs/315/

> I believe it would be acceptable to do nothing there.  In fact, pthread_kill with signal number zero does nothing. 

So, could possibly just check like  `#ifndef __ANDROID__` and do nothing or [use pthread_kill like here](https://github.com/axel-download-accelerator/axel/commit/411629ff2a7c8b1d896c62d0c09cd5b8e2e1ff67)

Not sure if I should spit this into its own issue, will continue trying to investigate when I have time

@anzhi0708 were you ever able to get around this or build on termux? Seems I'm following in your debugging as only way I got ocaml to build was by following https://github.com/ocaml/opam-repository/issues/22748

**Comment by seanbreckenridge** [comment]
Ok, it does at least build with this diff:

```diff
diff --git a/otherlibs/stdune/src/wait3_stubs.c b/otherlibs/stdune/src/wait3_stubs.c
index f84c9288d..edab23e02 100644
--- a/otherlibs/stdune/src/wait3_stubs.c
+++ b/otherlibs/stdune/src/wait3_stubs.c
@@ -45,6 +45,10 @@ static value alloc_process_status(int status) {
 
 static int wait_flag_table[] = {WNOHANG, WUNTRACED};
 
+pid_t wait3(int *status, int options, struct rusage *rusage) {
+  return wait4(-1, status, options, rusage);
+}
+
 value dune_wait3(value flags) {
   CAMLparam1(flags);
   CAMLlocal2(times, res);
diff --git a/vendor/spawn/src/spawn_stubs.c b/vendor/spawn/src/spawn_stubs.c
index fcb811726..d2338f084 100644
--- a/vendor/spawn/src/spawn_stubs.c
+++ b/vendor/spawn/src/spawn_stubs.c
@@ -703,7 +703,9 @@ CAMLprim value spawn_unix(value v_env,
      For instance:
      http://git.musl-libc.org/cgit/musl/tree/src/process/posix_spawn.c
   */
+  #if !defined(__ANDROID__)
   pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &cancel_state);
+  #endif
   sigfillset(&sigset);
   pthread_sigmask(SIG_SETMASK, &sigset, &saved_procmask);
 
@@ -750,7 +752,9 @@ CAMLprim value spawn_unix(value v_env,
 
   close(result_pipe[0]);
   pthread_sigmask(SIG_SETMASK, &saved_procmask, NULL);
+  #if !defined(__ANDROID__)
   pthread_setcancelstate(cancel_state, NULL);
+  #endif
 
   caml_leave_blocking_section();
 
```

```
$ make release                        
[ /data/data/com.termux/files/usr/tmp/Df1XXNU8O2/dune | main ] $ ./_build/default/bin/dune.exe                
dune: required COMMAND name is missing, must be one of 'build', 'cache', 'clean', 'coq', 'describe', 'diagnostics', 'exec', 'external-lib-deps', 'fmt', 'format-dune-file', 'help', 'init', 'install', 'installed-libraries', 'internal', 'monitor', 'ocaml', 'ocaml-merlin', 'pkg', 'printenv', 'promote', 'promotion', 'rpc', 'rules', 'runtest', 'show', 'shutdown', 'subst', 'test', 'top', 'uninstall', 'upgrade' or 'utop'.
Usage: dune COMMAND …
Try 'dune --help' for more information.
```

Am unsure if disabling the `pthread_setcancelstate` altogether is fine though, am not able to grasp the entirety of where this is used.

**Comment by Alizter** [comment]
@seanbreckenridge Are you able to do `./dune.exe build dune.install` and then `./dune.exe install`? The `./dune.exe` is the bootstrapped version of dune and `./dune.exe build dune.install` will build the actual version of dune you should use elsewhere.

**Comment by seanbreckenridge** [comment]
All but one task seems to compile:

```
$ ./dune.exe build dune.install         
Warning: Unable to read directory /data/data. Will not look for root in
parent directories.
Reason: /data/data: Permission denied
To remove this warning, set your root explicitly using --root.
Error: /data/data: Permission denied 
-> required by
   _build/default/bin/.main.eobjs/build_info__Build_info_data.ml-gen
-> required by
   _build/default/bin/.main.eobjs/native/build_info__Build_info_data.cmx
-> required by _build/default/bin/main.exe
-> required by _build/install/default/bin/dune
-> required by _build/default/dune.install
```

Could probably be configured with some environment variable but not sure whats failing.

Verbose output: https://gist.github.com/seanbreckenridge/737c672c93302caa0c2f0a5cb5937a0a

**Comment by Alizter** [comment]
Does it work when you set `--root` as `.`?

**Comment by seanbreckenridge** [comment]
Same error, the warning is gone though:

```
$ ./dune.exe build dune.install --root=.
Error: /data/data: Permission denied 
-> required by
   _build/default/bin/.main.eobjs/build_info__Build_info_data.ml-gen
-> required by
   _build/default/bin/.main.eobjs/native/build_info__Build_info_data.cmx
-> required by _build/default/bin/main.exe
-> required by _build/install/default/bin/dune
-> required by _build/default/dune.install
```

**Comment by Alizter** [comment]
The `--libdir` options all fall under `--prefix` however this is only relevant in the `dune install` command. That simply copies the files staged for installation that will be built when running `./dune.exe build dune.install`. It's strange that it is trying to access the root directory in termux however. I'm not quite sure what is going on there.

Let's try building the executable only. Try running:
```
./dune.exe build _build/install/default/bin/dune
```
if that works you can manually install that executable.

**Comment by Alizter** [comment]
I found a related issue which makes me think that this might be an OCaml issue. Can you try another version other than 5.0? 4.14 or 5.1 may be better.

**Comment by seanbreckenridge** [comment]
Same error on `4.14.1`:

```
[ /data/data/com.termux/files/usr/tmp/16LFx3sGQA/dune | main ] $ ocamlc --version
4.14.1
[ /data/data/com.termux/files/usr/tmp/16LFx3sGQA/dune | main ] $ opam --version  
2.0.4
[ /data/data/com.termux/files/usr/tmp/16LFx3sGQA/dune | main ] $ ./dune.exe build _build/install/default/bin/dune --root=.
Error: /data/data: Permission denied
-> required by
   _build/default/bin/.main.eobjs/build_info__Build_info_data.ml-gen
-> required by
   _build/default/bin/.main.eobjs/native/build_info__Build_info_data.cmx
-> required by _build/default/bin/main.exe
-> required by _build/install/default/bin/dune
```

**Comment by seanbreckenridge** [comment]
In any case at least one of the errors (just building `dune`) could possibly be solved by PR'ing the patch above? I think the `defined(__ANDROID__)` check if pretty normal, but I'm no C or android developer, this was all just googling on my end.  If someone more familiar with C/pthreads knows if just skipping the `pthread_setcancelstate` on android is fine, could make/test a PR (feel free to ping me).

Not sure if the `wait3` stub should be PR'd as well, or if the plan is to switch to `wait4`.

I've been able to build my project with dune, which was my original goal, but `opam install dune` still fails because of the missing `wait3`/`pthread` above and `./dune.exe build _build/install/default/bin/dune --root=.` failing.

Same error on 5.1.0, but I think this could be moved into its own issue, I dont think its related:

```
$ ocamlc --version
5.1.0
[ /data/data/com.termux/files/usr/tmp/16LFx3sGQA/dune | main ] $ ./dune.exe build _build/install/default/bin/dune --root=.
Error: /data/data: Permission denied
-> required by
   _build/default/bin/.main.eobjs/build_info__Build_info_data.ml-gen
-> required by
   _build/default/bin/.main.eobjs/native/build_info__Build_info_data.cmx
-> required by _build/default/bin/main.exe
-> required by _build/install/default/bin/dune
```

**Comment by rgrinberg** [comment]
Your patch to remove `pthread_setcancelstate` on Android is fine, but it should go into https://github.com/janestreet/spawn. We just pull from that library into the dune code periodically.

> Not sure if the wait3 stub should be PR'd as well, or if the plan is to switch to wait4.

Switching to `wait4` sounds fine to me. Although I would recommend that you rename the stubs file to `wait4` and expose the extra argument to make it a proper binding.

**Comment by seanbreckenridge** [comment]
Sounds good, will make a PR for the `pthread` code there.

By expose the extra argument for wait3 you mean [here in the proc.ml](https://github.com/ocaml/dune/blob/e8822c1d3e7bcbe0ad806fbaec7af3468b6e895b/otherlibs/stdune/src/proc.ml#L49-L52) and update the type in the `.mli` file, correct?

**Comment by rgrinberg** [comment]
> By expose the extra argument for wait3 you mean [here in the proc.ml](https://github.com/ocaml/dune/blob/e8822c1d3e7bcbe0ad806fbaec7af3468b6e895b/otherlibs/stdune/src/proc.ml#L49-L52) and update the type in the .mli file, correct?

Yes, that's right. Essentially remove all mentions of `wait3` in the codebase.

**Comment by seanbreckenridge** [comment]
Hm, actually looking at <https://man7.org/linux/man-pages/man2/wait4.2.html>

> These functions are nonstandard; in new programs, the use of [waitpid(2)](https://man7.org/linux/man-pages/man2/waitpid.2.html) or [waitid(2)](https://man7.org/linux/man-pages/man2/waitid.2.html) is preferable

> The wait3() and wait4() system calls are similar to [waitpid(2)](https://man7.org/linux/man-pages/man2/waitpid.2.html),
       but additionally return resource usage information about the
       child in the structure pointed to by rusage.

but. this is not a new program, and we do return the `rusage`  I suppose we can continue to use `wait4`?

Have the `wait4` stubs working, but am glad to switch to `waitpid` if we'd want to instead

**Comment by seanbreckenridge** [comment]
Alright, did a test build by pulling in the new janestreet/spawn stubs, but am blocked by https://github.com/janestreet/spawn/pull/48 now

```
git clone https://github.com/ocaml/dune
git clone https://github.com/janestreet/spawn
rsync -Pavh ./spawn/src/ ./dune/vendor/spawn/src/

make release
```

````
$ make release
ocamlc -output-complete-exe -w -24 -g -o .duneboot.exe -I boot -I +unix unix.cma boot/libs.ml boot/duneboot.ml
./.duneboot.exe
Done: 146/1349 (jobs: 8)cd _boot && /data/data/com.termux/files/home/.opam/5.1.0/bin/ocamlopt.opt -c -g -I +unix -I +threads spawn_stubs.c
vendor/spawn/src/spawn_stubs.c:492:20: error: call to undeclared function 'caml_convert_signal_number'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
      int signal = caml_convert_signal_number(Long_val(Field(v_signals_list, 0)));
                   ^
1 error generated.
```

Am willing to continue testing this once thats fixed, feel free to ping me

**Comment by Alizter** [comment]
I was able to repro your issue with the permissions and I passed `--debug-backtrace` to get a better idea what was going on:
![image](https://github.com/ocaml/dune/assets/8614547/d174e726-339d-46d3-8f4d-21276fd0ef93)

@rgrinberg Looks like two things are broken here:
1. vcs stuff is looking in dirs it shouldn't be.
2. The root detection doesn't seem to work.

**Comment by seanbreckenridge** [comment]
may be related, I've been running all these commands in my `tmp` directory, which may have some impact?

```
[ ~ ] $ cdtmp      
[ /data/data/com.termux/files/usr/tmp/T7i3cJMnD0 ] $ git clone -q https://github.com/ocaml/dune && cd dune; make release

ocamlc -output-complete-exe -w -24 -g -o .duneboot.exe -I boot -I +unix unix.cma boot/libs.ml boot/duneboot.ml
./.duneboot.exe
Done: 147/1349 (jobs: 8)cd _boot && /data/data/com.termux/files/home/.opam/5.1.0/bin/ocamlopt.opt -c -g -I +unix -I +threads spawn_stubs.c
vendor/spawn/src/spawn_stubs.c:706:3: error: call to undeclared function 'pthread_setcancelstate'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &cancel_state);
  ^
```

(fails with the expected error, in `/data/data/com.termux/files/usr/tmp`)

This is how I've been testing all the builds here, because when I originally tried to do it in my home directory (e.g. in my Downloads folder at  `/data/data/com.termux/files/usr/files/home/Downloads`):

It fails with a different error:

```
$ pwd
/data/data/com.termux/files/home/Downloads/dune

[ ~/Downloads ] $ git clone -q https://github.com/ocaml/dune && cd dune && make release  
ocamlc -output-complete-exe -w -24 -g -o .duneboot.exe -I boot -I +unix unix.cma boot/libs.ml boot/duneboot.ml
./.duneboot.exe
sh: ./.duneboot.exe: can't execute: Permission denied
make: *** [Makefile:53: _boot/dune.exe] Error 126
```

I get the same output just running `ocaml boot/bootstrap.ml`

**Comment by romanovj** [comment]
> Ok, it does build with this diff:
> 
> ```diff
> diff --git a/otherlibs/stdune/src/wait3_stubs.c b/otherlibs/stdune/src/wait3_stubs.c
> index f84c9288d..edab23e02 100644
> --- a/otherlibs/stdune/src/wait3_stubs.c
> +++ b/otherlibs/stdune/src/wait3_stubs.c
> @@ -45,6 +45,10 @@ static value alloc_process_status(int status) {
>  
>  static int wait_flag_table[] = {WNOHANG, WUNTRACED};
>  
> +pid_t wait3(int *status, int options, struct rusage *rusage) {
> +  return wait4(-1, status, options, rusage);
> +}
> +
>  value dune_wait3(value flags) {
>    CAMLparam1(flags);
>    CAMLlocal2(times, res);
> diff --git a/vendor/spawn/src/spawn_stubs.c b/vendor/spawn/src/spawn_stubs.c
> index fcb811726..d2338f084 100644
> --- a/vendor/spawn/src/spawn_stubs.c
> +++ b/vendor/spawn/src/spawn_stubs.c
> @@ -703,7 +703,9 @@ CAMLprim value spawn_unix(value v_env,
>       For instance:
>       http://git.musl-libc.org/cgit/musl/tree/src/process/posix_spawn.c
>    */
> +  #if !defined(__ANDROID__)
>    pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &cancel_state);
> +  #endif
>    sigfillset(&sigset);
>    pthread_sigmask(SIG_SETMASK, &sigset, &saved_procmask);
>  
> @@ -750,7 +752,9 @@ CAMLprim value spawn_unix(value v_env,
>  
>    close(result_pipe[0]);
>    pthread_sigmask(SIG_SETMASK, &saved_procmask, NULL);
> +  #if !defined(__ANDROID__)
>    pthread_setcancelstate(cancel_state, NULL);
> +  #endif
>  
>    caml_leave_blocking_section();
>  
> ```
> 
> ```
> $ make release                        
> [ /data/data/com.termux/files/usr/tmp/Df1XXNU8O2/dune | main ] $ ./_build/default/bin/dune.exe                
> dune: required COMMAND name is missing, must be one of 'build', 'cache', 'clean', 'coq', 'describe', 'diagnostics', 'exec', 'external-lib-deps', 'fmt', 'format-dune-file', 'help', 'init', 'install', 'installed-libraries', 'internal', 'monitor', 'ocaml', 'ocaml-merlin', 'pkg', 'printenv', 'promote', 'promotion', 'rpc', 'rules', 'runtest', 'show', 'shutdown', 'subst', 'test', 'top', 'uninstall', 'upgrade' or 'utop'.
> Usage: dune COMMAND …
> Try 'dune --help' for more information.
> ```
> 
> I then just copied `_build/default/bin/dune.exe` onto my `$PATH` since `make install` fails; tried setting the `$PREFIX` but didnt seem to fix it:
> 
> ```
> $ PREFIX="${HOME}/../usr/" make install
> ./_boot/dune.exe install --prefix /data/data/com.termux/files/usr   dune
> Warning: Unable to read directory /data/data. Will not look for root in
> parent directories.
> Reason: /data/data: Permission denied
> To remove this warning, set your root explicitly using --root.
> Error: /data/data: Permission denied
> make: *** [Makefile:63: install] Error 1
> ```
> 
> Am unsure if disabling the `pthread_setcancelstate` altogether is fine though, am not able to grasp the entirety of where this is used.
> 
> Am able to successfully build [my project which I was hoping to](https://github.com/seanbreckenridge/calc_ans) on termux

termux-chroot
cd dune
make install PREFIX=$PREFIX

**Comment by romanovj** [comment]
> may be related, I've been running all these commands in my `tmp` directory, which may have some impact?
> 
> Trying to build in the tmp dir:
> 
> ```
> [ ~ ] $ cdtmp      
> [ /data/data/com.termux/files/usr/tmp/T7i3cJMnD0 ] $ git clone -q https://github.com/ocaml/dune && cd dune; make release
> 
> ocamlc -output-complete-exe -w -24 -g -o .duneboot.exe -I boot -I +unix unix.cma boot/libs.ml boot/duneboot.ml
> ./.duneboot.exe
> Done: 147/1349 (jobs: 8)cd _boot && /data/data/com.termux/files/home/.opam/5.1.0/bin/ocamlopt.opt -c -g -I +unix -I +threads spawn_stubs.c
> vendor/spawn/src/spawn_stubs.c:706:3: error: call to undeclared function 'pthread_setcancelstate'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
>   pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &cancel_state);
>   ^
> ```
> 
> (fails with the expected error, in `/data/data/com.termux/files/usr/tmp`)
> 
> This is how I've been testing all the builds here, because when I originally tried to do it in my home directory (e.g. in my Downloads folder at `/data/data/com.termux/files/usr/files/home/Downloads`), it fails with a different error:
> 
> ```
> $ pwd
> /data/data/com.termux/files/home/Downloads
> 
> [ ~/Downloads ] $ git clone -q https://github.com/ocaml/dune && cd dune && make release  
> ocamlc -output-complete-exe -w -24 -g -o .duneboot.exe -I boot -I +unix unix.cma boot/libs.ml boot/duneboot.ml
> ./.duneboot.exe
> sh: ./.duneboot.exe: can't execute: Permission denied
> make: *** [Makefile:53: _boot/dune.exe] Error 126
> ```
> 
> I get the same output just running `ocaml boot/bootstrap.ml`

/data/data/com.termux/files/usr/files/home/Downloads
should be softlink to external storage which mounts with noexec

**Comment by romanovj** [comment]
> Alright, did a test build by pulling in the new janestreet/spawn stubs, but doesnt build because of [janestreet/spawn#48](https://github.com/janestreet/spawn/pull/48) now
> 
> ```
> git clone https://github.com/ocaml/dune
> git clone https://github.com/janestreet/spawn
> rsync -Pavh ./spawn/src/ ./dune/vendor/spawn/src/
> 
> cd dune
> 
> make release
> ```
> 
> ```
> $ make release
> ocamlc -output-complete-exe -w -24 -g -o .duneboot.exe -I boot -I +unix unix.cma boot/libs.ml boot/duneboot.ml
> ./.duneboot.exe
> Done: 146/1349 (jobs: 8)cd _boot && /data/data/com.termux/files/home/.opam/5.1.0/bin/ocamlopt.opt -c -g -I +unix -I +threads spawn_stubs.c
> vendor/spawn/src/spawn_stubs.c:492:20: error: call to undeclared function 'caml_convert_signal_number'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
>       int signal = caml_convert_signal_number(Long_val(Field(v_signals_list, 0)));
>                    ^
> 1 error generated.
> ```
> 
> Am willing to continue testing this once thats fixed, feel free to ping me
> 
> Am still able to build dune by following something similar to [here](https://github.com/ocaml/dune/issues/8676#issuecomment-1722574714), just not able to when trying to build against fresh main/spawn stubs
> 
> (Also, still fails to run `./dune.exe build _build/install/default/bin/dune --root=.`, so this likely wont be an easy `opam install dune` even after thats fixed - that could be moved to a separate issue possibly)

vendor/spawn/src/spawn_stubs.c

imports caml/memory.h which imports signals.h

after it this construction isn't working at least in termux's clang

```
#define CAML_INTERNALS
/* for [caml_convert_signal_number] */
#include <caml/signals.h>
#undef CAML_INTERNALS
```

add 
CAMLextern int caml_convert_signal_number(int);
before
int signal = caml_convert_signal_number(Long_val(Field(v_signals_list, 0)));

**Comment by seanbreckenridge** [comment]
Ok, yeah, added the CAMLextern line:

```
/* for [caml_convert_signal_number] */
#include <caml/signals.h>
CAMLextern int caml_convert_signal_number(int);
#undef CAML_INTERNALS
```

Doing `make release` I did get a:

```
3 |  (public_name spawn)
                  ^^^^^
Error: The current scope doesn't define package "spawn".
```

But I just removed the `public_name` list from the dune file, as it wasnt there previously (I expect this'll be fixed when new stubs are merged in)

Did something like:

```shell
pkg install proot  #  for termux-chroot
cd ~  # just going to do it in my home dir
mkdir tmp
cd tmp
termux-chroot

git clone --depth=1 https://github.com/ocaml/dune

# temp fixes for spawn
git clone https://github.com/janestreet/spawn
rsync -Pavh ./spawn/src/ ./dune/vendor/spawn/src/
# remove public_name from the dune file
# add the CAMLextern line

cd dune
unset LD_PRELOAD  # https://github.com/termux/proot/issues/74
make release
make install PREFIX=$PREFIX
```

And I get no errors :+1:  

```
[ /data/data/com.termux/files/home/tmp/dune | main ] $ which dune
/data/data/com.termux/files/usr/bin/dune
[ /data/data/com.termux/files/home/tmp/dune | main ] $ ldd $(which dune)
	libandroid-shmem.so => /data/data/com.termux/files/usr/lib/libandroid-shmem.so
	libzstd.so.1 => /data/data/com.termux/files/usr/lib/libzstd.so.1
	libm.so => /system/lib64/libm.so
	libc.so => /system/lib64/libc.so
	libdl.so => /system/lib64/libdl.so
	liblog.so => /system/lib64/liblog.so
	libc++.so => /system/lib64/libc++.so
	ld-android.so => /system/lib64/ld-android.so
```

I dont expect `dune` to just build with no changes as is often the case you have to do something custom on termux, but this could at least be documented once the spawn stubs have been merged in.

**Comment by emillon** [comment]
Ok, I'll pull the spawn changes.

**Comment by rgrinberg** [comment]
Heads up: the changes are currently in my personal fork of spawn.

You could create a fork of spawn in ocaml-dune and apply my changes on top of master of janestreet/spawn

**Comment by seanbreckenridge** [comment]
Nice, since #8768, can do this with just

After installing `proot`:

```bash
unset LD_PRELOAD
termux-chroot

git clone --depth=1 https://github.com/ocaml/dune
cd dune

make release
make install PREFIX=$PREFIX
```

Perhaps could be documented somewhere, not sure

I think this is good enough to close this issue, not going to get much easier that on termux

**Comment by Alizter** [comment]
I'm also able to build on termux now. Thanks!
